src/BlockTypes/CollectionFilters.php

```diff
@@ -198,4 +203,71 @@ private function get_inner_blocks_recursive( $block, $results = array() ) {
 		}
 		return $results;
 	}
+
+	/**
+	 * Get formatted products params for ProductCollectionData route from the
+	 * query context.
+	 *
+	 * @param array $query The query context.
+	 * @return array
+	 */
+	private function get_formatted_products_params( $query ) {
+		$params = array();
+
+		if ( empty( $query['isProductCollectionBlock'] ) ) {
+			return $params;
+		}
+
+		// Params that doesn't require additional conversion.
+		$params_mapping = array(
+			'exclude'                       => 'exclude',
+			'offset'                        => 'offset',
+			'order'                         => 'order',
+			'orderBy'                       => 'orderby',
+			'pages'                         => 'page',
+			'parents'                       => 'parent',
+			'perPage'                       => 'per_page',
+			'search'                        => 'search',
+			'woocommerceStockStatus'        => 'stock_status',
+			'woocommerceOnSale'             => 'on_sale',
+			'woocommerceHandPickedProducts' => 'include',
+		);
+
+		foreach ( $query as $key => $value ) {
+			if ( isset( $params_mapping[ $key ] ) ) {
+				$params[ $params_mapping[ $key ] ] = $value;
+			}
+		}
+
+		if ( ! empty( $query['taxQuery'] ) ) {
+			foreach ( $query['taxQuery'] as $taxonomy => $value ) {
```

Comment:
Can you share an example in JS on the idea you're talking about? I want to make sure we're on the same page. We're doing the similar thing for editor anyway so it won't be thrown away.

> so that I could iterate the keys in the same loop above and do all transformations there.

Do you mean processing top level keys and nested keys inside one loop? Also process the transformation for both value can passed as it is and value need different transformation?

> Then, I would refactor these little transformations into their own functions as to put away implementation details from this method.

Please keep in mind that the current method is already implementation details. I'm happy to split it down btw, as long as it doesn't make the function harder to read.

Again, I think it's easier for us if you can provide some code demonstrate your idea.

Reply:
> Can you share an example in JS on the idea you're talking about? I want to make sure we're on the same page. We're doing the similar thing for editor anyway so it won't be thrown away.

Maybe something like this?

```ts
const PARAMS_MAP = {
  parents: "parent",
  perPage: "per_page",
  taxQuery: ( key ) => {
    const map = {
         product_cat: "category",
         product_tag: "tag",
    };
    
    return map[ key ] || `_unstable_tax_${ key }`;
  }
}

return Object.entries( query ).reduce( ( acc, [ key, value ] ) => {
   const mapper = PARAMS_MAP[ key ];
   const newKey = typeof mapper === "function" ? mapper( key ) : mapper;
   
   acc[ newKey ] = value;
   return acc;
}, {} );
```

> Do you mean processing top level keys and nested keys inside one loop? Also process the transformation for both value can passed as it is and value need different transformation?

Sorry not sure what you mean with this?

> Please keep in mind that the current method is already implementation details. I'm happy to split it down btw, as long as it doesn't make the function harder to read.

Yes, that's a good point. Maybe I'm too trigger-happy when it comes to split stuff. But I like the idea of having the key/value conversion responsibility to be separate. But maybe it's too overkill!

Reply:
> > Do you mean processing top level keys and nested keys inside one loop?

> Sorry not sure what you mean with this?

In the code above, you process keys of both top level param like `parents`, `perPage`, and nested keys like `product_cat` and `product_tag`. If we have like 5 or 10 params like `taxQuery`, that probably makes sense to map everything in one place, but here we're doing so for just `taxQuery`.

> > Also process the transformation for both value can passed as it is and value need different transformation?

We have to transform the value of nested keys of `taxQuery`, they're originally arrays, but we need comma separated strings to pass it to the Store API.

Btw, just a note if we reuse the snippet above, I tested it and it doesn't return the desired shape.

<details>
<summary>Screenshot</summary>

<img width="719" alt="image" src="https://github.com/woocommerce/woocommerce-blocks/assets/5423135/eca12dc4-eecc-41eb-be22-a59e31e74b5b">

</details>